python/cuml/test/test_target_encoder.py

```diff
@@ -0,0 +1,48 @@
+# Copyright (c) 2019, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from cuml.preprocessing.TargetEncoder import TargetEncoder
+import cudf
+import numpy as np
+from cuml.test.utils import array_equal
+
+
+def test_targetencoder_fit_transform():
+    train = cudf.DataFrame({'category': ['a', 'b', 'b', 'a'],
```

Comment:
Would be great to beef up the testing... I think a couple of scenarios to try are:

- only one category present
- newly-encountered categories in test
- input from cupy array and other datatypes
- other datatypes for X -- do we support integer values for the categorical input?
- high cardinality example

hopefully these can mostly become parametrizations?

Reply:
How to check the correctness of the high cardinality example? It typically needs a cpu baseline but I'm not aware of such baseline.

Reply:
Ok, `cumlArray` doesn't support string columns right now. I'll do it manually.

Reply:
To test correctness, I'd build an example with a known mean per target. What about something like:
```
expected_outputs = {}
y_list = ...
x_list = ...

for i in range(20):
  card_val = 'f%d' % i
  n_samples = i+1
  for j in range(n_samples):
    x_list.append(card_val)
    y_list.append(i*10.0 + j)
  expected_val[card_val] = i*10.0 + (n_samples-1)/2.0 + ... however to handle smoothing term ...
```
Then you can parametrize that however you want for cardinality / size / etc.

Reply:
Thank you. I added a large random test which is equivalent to your approach. However, this kind of check only works for `transform` test data. For the `fit_transform` result, it is not this predefined per level target. The encoding result also depends on how train data is split to `k` folds. So it has to be as complex as the `TargetEncoder.fit_transform` to check the results. I leave it out for now.

Reply:
I'm pretty sure the implementation is correct based on the walkthrough notebook.